File: backend/src/services/companion.service.ts
import { promises as fs } from 'fs';
import path from 'path';
import { Companion } from '../models/companion';

/**
 * Loads the companion catalogue from companions.json.
 * The file is read once and cached for the process lifetime.
 */
export class CompanionService {
  private readonly filePath = path.resolve(__dirname, '../../data/static/companions.json');
  private cache: Companion[] | null = null;

  async getAll(): Promise<Companion[]> {
    if (this.cache) {
      return this.cache;
    }

    const raw = await fs.readFile(this.filePath, 'utf8');
    const parsed = JSON.parse(raw) as Companion[];
    this.cache = parsed;
    return parsed;
  }
}

File: backend-csharp/TheGameIsOn.API/Services/CompanionService.cs
using System.Text.Json;
using TheGameIsOn.API.Helpers;
using TheGameIsOn.API.Models;

namespace TheGameIsOn.API.Services;

public interface ICompanionService
{
    Task<List<Companion>> GetAllAsync();
}

/// <summary>
/// Loads the companion catalogue from companions.json.
/// The file is read once and cached for the process lifetime.
/// </summary>
public class CompanionService : ICompanionService
{
    private readonly string _filePath;
    private List<Companion>? _cache;
    private readonly SemaphoreSlim _lock = new(1, 1);

    public CompanionService(IConfiguration configuration)
    {
        var root = configuration["DataRoot"] ?? "data";
        _filePath = Path.Combine(root, "static", "companions.json");
    }

    public async Task<List<Companion>> GetAllAsync()
    {
        if (_cache is not null) return _cache;

        await _lock.WaitAsync();
        try
        {
            if (_cache is not null) return _cache;

            List<Companion> companions;
            try
            {
                var raw = await File.ReadAllTextAsync(_filePath);
                companions = JsonSerializer.Deserialize<List<Companion>>(raw, JsonDefaults.Options) ?? new();
            }
            catch
            {
                companions = new();
            }

            _cache = companions;
            return _cache;
        }
        finally
        {
            _lock.Release();
        }
    }
}
